mvp-spec/partials-updater/src/lexer.ts

```diff
@@ -0,0 +1,99 @@
+export const PartialBeginStartKey = "<!--partial-begin"
+export const PartialBeginEndKey = "-->\r\n"
+export const PartialEndKey = "<!--partial-end-->\r\n"
+
+/** All type of tokens for a simplified lexing for partials. */
+export enum LexerTokenType {
+    PlainText = "TEXT",
+    PartialBegin = "PARTIAL_BEGIN",
+    PartialEnd = "PARTIAL_END"
+}
+
+/** Token describing the different part of a document. */
+export interface LexerToken {
+    type: LexerTokenType;
+    text: string;
+}
+
+/** Split a content in tokens. */
+export function lex(content: string): LexerToken[] {
+    const tokens = new Array<LexerToken>();
+    const lines = getLines(content);
+    let textAcc = '';
+    lines.forEach(line => {
+        if (line.startsWith(PartialBeginStartKey) && line.endsWith(PartialBeginEndKey)) {
+            if (textAcc != '') {
+                const token: LexerToken = {
+                    type: LexerTokenType.PlainText,
+                    text: textAcc
+                };
+                tokens.push(token);
+                textAcc = '';
+            }
+            const token: LexerToken = {
+                type: LexerTokenType.PartialBegin,
+                text: line
+            };
+            tokens.push(token);
+        } else if (line == PartialEndKey) {
+            if (textAcc != '') {
+                const token: LexerToken = {
+                    type: LexerTokenType.PlainText,
+                    text: textAcc
+                };
+                tokens.push(token);
+                textAcc ='';
+            }
+            const token: LexerToken = {
+                type: LexerTokenType.PartialEnd,
+                text: line
+            }
+            tokens.push(token);
+        } else {
+            textAcc += line;
+        }
+    });
+    if (textAcc != '') {
+        const token: LexerToken = {
+            type: LexerTokenType.PlainText,
+            text: textAcc
+        };
+        tokens.push(token);
+        textAcc ='';
+    }
+    return tokens;
+}
+
+/**
+ * Split a content in lines including the separators.
+ * 
+ * @param content String to split in lines
+ * @returns An array of lines including the separator.
+ */
+function getLines(content: string): string[] {
+    const separators = [ '\n', '\r\n' ];
```

Comment:
What about for this entire function:
`return content.match(/[^\r\n]+(\r?\n)?|\r?\n/g);`
?

Reply:
I don't push for regular expressions in all cases.
In this case to be honest, I find it 10x more readable ;-)

Reply:
To say it differently, I think this regexp is relatively simple and **much less error-prone** than 25 lines of code with nested loops and `break` statement 😅